modules/apps/portal/portal-instance-lifecycle-impl/src/main/java/com/liferay/portal/instance/lifecycle/internal/PortalInstanceLifecycleListenerManagerImpl.java

```diff
@@ -291,12 +236,101 @@ protected void unregisterCompany(
 		TransactionConfig.Factory.create(
 			Propagation.REQUIRED, new Class<?>[] {Exception.class});
 
+	private BundleContext _bundleContext;
 	private final Set<Company> _companies = new CopyOnWriteArraySet<>();
 
 	@Reference
 	private CompanyLocalService _companyLocalService;
 
-	private final Set<PortalInstanceLifecycleListener>
-		_portalInstanceLifecycleListeners = new CopyOnWriteArraySet<>();
+	private ServiceTrackerList<PortalInstanceLifecycleListener>
+		_serviceTrackerList;
+
+	private class PortalInstanceLifecycleListenerServiceTrackerCustomizer
+		implements ServiceTrackerCustomizer
+			<PortalInstanceLifecycleListener, PortalInstanceLifecycleListener> {
+
+		@Override
+		public PortalInstanceLifecycleListener addingService(
+			ServiceReference<PortalInstanceLifecycleListener>
+				serviceReference) {
+
+			PortalInstanceLifecycleListener portalInstanceLifecycleListener =
+				_bundleContext.getService(serviceReference);
+
+			if (_companies.isEmpty()) {
+				return portalInstanceLifecycleListener;
+			}
+
+			Iterator<Company> iterator = _companies.iterator();
+
+			while (iterator.hasNext()) {
+				Company company = iterator.next();
+
+				Company fetchedCompany = _companyLocalService.fetchCompanyById(
+					company.getCompanyId());
+
+				if (fetchedCompany == null) {
+					try {
+						TransactionInvokerUtil.invoke(
+							_transactionConfig,
+							() -> {
+								unregisterCompany(company);
+
+								return portalInstanceLifecycleListener;
+							});
+					}
+					catch (Throwable throwable) {
+						throw new RuntimeException(throwable);
+					}
+				}
+			}
```

Comment:
@W2jx, if catch happened, the service won‘t be added to _serviceTrackerList. This is different from the previous logic. I think you may not throw the Exception, instead of using _log.error(), and then return portalInstanceLifecycleListener.